apps/internal/base/internal/storage/partitioned_storage.go

```diff
@@ -41,35 +41,47 @@ func (m *PartitionedManager) Read(ctx context.Context, authParameters authority.
 	clientID := authParameters.ClientID
 	scopes := authParameters.Scopes
 
-	metadata, err := m.getMetadataEntry(ctx, authParameters.AuthorityInfo)
-	if err != nil {
-		return TokenResponse{}, err
+	// fetch metadata if authority isn't explicitly trusted or instanceDiscovery is false
+	aliases := authParameters.KnownAuthorityHosts
+	var knownAuthority bool = false
+	for _, alias := range aliases {
+		if alias == authParameters.AuthorityInfo.Host {
+			knownAuthority = true
+			break
+		}
+	}
+	if !knownAuthority && !authParameters.AuthorityInfo.InstanceDiscoveryDisabled {
+		metadata, err := m.getMetadataEntry(ctx, authParameters.AuthorityInfo)
+		if err != nil {
+			return TokenResponse{}, err
+		}
+		aliases = metadata.Aliases
```

Comment:
I believe this could be much simpler. Functionally, `aliases` are the hosts considered to match this token request. The requested host, `authParameters.AuthorityInfo.Host`, is of course always a match. We do need to know any aliases of that host to prevent unnecessary cache misses. The client should discover these from instance metadata, but only if instance discovery is enabled. How about this then:
```go
// aliases are the hosts matching this auth attempt. The requested host always matches. If
// instance discovery is permitted, request instance metadata to discover any other aliases.
aliases := []string{authParameters.AuthorityInfo.Host}
if !authParameters.AuthorityInfo.InstanceDiscoveryDisabled {
	metadata, err := m.getMetadataEntry(ctx, authParameters.AuthorityInfo)
	if err != nil {
		return TokenResponse{}, err
	}
	aliases = metadata.Aliases
}
```
With that, the caller doesn't need to set `KnownAuthorities`. If we also replace `WithKnownAuthorityHosts` with `WithInstanceDiscovery(false)` [here](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/16c062c187ac55067ece02de9fe65f2f7b4033cd/apps/confidential/confidential.go#L396), we no longer need the field or its setter, which is a win because that API is a vestige of a rejected design.